code/modules/modular_computers/file_system/programs/generic/ntdownloader.dm

```diff
@@ -0,0 +1,247 @@
+	if(..())
+		return 1
+	if(href_list["PRG_downloadfile"])
+		if(!downloaded_file)
+			begin_file_download(href_list["PRG_downloadfile"], usr.stats.getStat(STAT_COG))
+		else if(check_file_download(href_list["PRG_downloadfile"]) && !downloads_queue.Find(href_list["PRG_downloadfile"]) && downloaded_file.filename != href_list["PRG_downloadfile"])
+			downloads_queue[href_list["PRG_downloadfile"]] = usr.stats.getStat(STAT_COG)
+		return 1
+	if(href_list["PRG_removequeued"])
+		downloads_queue.Remove(href_list["PRG_removequeued"])
+		return 1
+	if(href_list["PRG_reseterror"])
+		if(downloaderror)
+			download_completion = 0
+			download_netspeed = 0
+			downloaded_file = null
+			downloaderror = ""
+		return 1
+	if(href_list["download_pause"])
+		download_paused = !download_paused
+		if (download_paused)
+			ui_header = "downloader_paused.gif"
+		else
+			ui_header = "downloader_running.gif"
+		return 1
+
+	if(href_list["download_stop"])
+		if(downloaded_file)
+			abort_file_download()
+		return 1
+	return 0
+
+/datum/nano_module/program/computer_ntnetdownload
+	name = "Network Downloader"
+	var/obj/item/modular_computer/my_computer = null
+
+/datum/nano_module/program/computer_ntnetdownload/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1, var/datum/topic_state/state = GLOB.default_state)
+	if(program)
+		my_computer = program.computer
+
+	if(!istype(my_computer))
+		return
+
+	var/list/data = list()
+	var/datum/computer_file/program/ntnetdownload/prog = program
+	// For now limited to execution by the downloader program
+	if(!prog || !istype(prog))
+		return
+	if(program)
+		data = program.get_header_data()
+
+	// This IF cuts on data transferred to client, so i guess it's worth it.
+	if(prog.downloaderror) // Download errored. Wait until user resets the program.
+		data["error"] = prog.downloaderror
+	if(prog.downloaded_file) // Download running. Wait please..
+		data["downloadname"] = prog.downloaded_file.filename
+		data["downloaddesc"] = prog.downloaded_file.filedesc
+		data["downloadsize"] = prog.downloaded_file.size
+		data["downloadspeed"] = prog.download_netspeed
+		data["downloadcompletion"] = round(prog.download_completion, 0.1)
+
+	data["download_paused"] = prog.download_paused
+	data["disk_size"] = my_computer.hard_drive.max_capacity
+	data["disk_used"] = my_computer.hard_drive.used_capacity
+
+	var/obj/item/weapon/computer_hardware/hard_drive/HDD = program.computer.hard_drive
+	if(!HDD)
+		return 1
+	var/list/datum/computer_file/program/installed_programs = list()
+	for(var/datum/computer_file/program/P in HDD.stored_files)
+		installed_programs.Add(P)
+
+	var/list/datum/computer_file/program/downloadable_programs = list()
+	for(var/datum/computer_file/program/P in ntnet_global.available_station_software)
+		downloadable_programs.Add(P)
+
+	for(var/datum/computer_file/program/P in downloadable_programs) //Removeing programs from list that are already present on HDD
```

Comment:
Here's code to not display things that are already installed, does it work? Did you add it?

Reply:
Yes i did, and it works

Reply:
You can track my changes in commits, especially now cuz it easier